reintegrating most of NCBI's FreeTDS changes

ucko opened the pull request:
Per the mailing list discussion I recently started, I've gone ahead and published my fork.  A few quick notes:
* My initial branch (complete with one patch I cut out) is available as `ncbi-merge-request-1pre1`.
* Timestamps had been a mess and were always artificial (the changes having been ported over from a Subversion repository, often with some regrouping), so I reset them all at the start of June.
* When multiple developers made appreciable contributions to changes, I've used the `Co-Authored-By:` convention to credit secondary authors.
* Some contributors are no longer at NCBI, but I listed them under the email addresses they had at the time (timestamp resets notwithstanding).

I'm open to reworking or regrouping these changes as needed.

freddy77 commented on include/freetds/thread.h:
```diff
@@ -353,6 +353,14 @@ static inline int tds_mutex_trylock(tds_mutex *mtx)
 	ret = tds_raw_mutex_trylock(&mtx->mtx);
 	if (!ret) {
+#ifdef _WIN32
+		/* disallow recursive locking! */
```
This part of code affects only debug build, is this expected?
In this case I would just assert so to catch the issues during testing.

freddy77 replied:
Now on Windows mutexes also are not recursive so this patch is not necessary anymore.

ucko replied:
OK, thanks!

ucko commented:
Fair question.  Without this patch to freetds/thread.h, tds/unittests/corrupt and dblib/unittests/cancel were both failing in Windows debug builds, IIRC with assertion errors from tds_mutex_unlock that naturally can’t be an issue in release builds.  With this patch, the code that tries to use a lock recursively in these contexts recovers from this non-fatal error, as it already needs to do with POSIX mutexes.  Sorry for not making that clearer in the log message!

freddy77 commented on src/tds/net.c:
```diff
@@ -760,7 +760,7 @@ tds_socket_write(TDSCONNECTION *conn, TDSSOCKET *tds, const unsigned char *buf,
 #if ENABLE_EXTRA_CHECKS
 	/* this simulate the fact that send can return less bytes */
-	if (buflen >= 5) {
+	if (buflen >= 11) {
```
Merged the patch.

ucko replied:
Thanks!  I'm all for exercising code paths that might not otherwise occur often in practice, but not in ways that potentially break communication.

freddy77 commented on src/odbc/odbc.c:
```diff
@@ -1329,8 +1329,8 @@ SQLSetEnvAttr(SQLHENV henv, SQLINTEGER Attribute, SQLPOINTER Value, SQLINTEGER S
 		/* TODO - Make this really work */
-		/* env->attr.output_nts = i_val; */
-		env->attr.output_nts = SQL_TRUE;
+		env->attr.output_nts = i_val;
```
I cannot see any tests that prove that is safe to do it.

ucko replied:
OK, yeah, this change makes no difference to your tests, and we don't otherwise use the ODBC library these days (having settled on the CT-LIB API).  I'll drop it.

freddy77 commented on src/dblib/unittests/numeric.c:
```diff
@@ -172,7 +172,7 @@ main(int argc, char **argv)
 	read_login_info(argc, argv);
 
-	dbsetversion(DBVERSION_100);
+	/* dbsetversion(DBVERSION_100); */
```
FreeTDS is tested with success against MSSQL, can you elaborate the problem?
I think this is here to make the test work using Sybase libraries.

ucko replied:
With cb97ca1 merged, this `dbsetversion` call causes FreeTDS to insist on using TDS 5.0, which only Sybase supports.  I suppose I should squash those changes into a single patch.

freddy77 commented on src/odbc/connectparams.c:
```diff
@@ -398,7 +398,8 @@ odbc_parse_connect_string(TDS_ERRS *errs, const char *connect_string, const char
 		} else if (CHK_PARAM(PacketSize)) {
 			tds_parse_conf_section(TDS_STR_BLKSZ, tds_dstr_cstr(&value), login);
-		} else if (CHK_PARAM(ClientCharset)) {
+		} else if (CHK_PARAM(ClientCharset)
```
Merged with this addition and documentation too.

ucko replied:
Thanks!

freddy77 commented on src/odbc/winsetup.c:
```diff
@@ -423,16 +424,17 @@ ConfigTranslator(HWND hwndParent, DWORD * pvOption)
 {
-	TCHAR fn[MAX_PATH], full_fn[MAX_PATH];
-	LPTSTR name;
+	char fn[MAX_PATH], full_fn[MAX_PATH];
```
Updated code to use always WCHAR in this function so driver can be installed on any directory.

ucko replied:
Even better, thanks!

freddy77 commented on win32/initnet.c:
```diff
@@ -4,12 +4,13 @@
 #include <windows.h>
 
-#ifdef DLL_EXPORT
-
```
this is bogus, winlogin.c will compile finding the variable but will fail running as the resources (rc file) needed is not included.
Better to fail compiling and fixing the project files.

ucko replied:
I'll take another look, thanks.

freddy77 commented on src/tds/login.c:
```diff
@@ -908,7 +911,7 @@ tds7_send_login(TDSSOCKET * tds, const TDSLOGIN * login)
 #define SET_FIELD_DSTR(field, dstr) do { \
 	data_fields[field].ptr = tds_dstr_cstr(&(dstr)); \
-	data_fields[field].len = tds_dstr_len(&(dstr)); \
+	data_fields[field].len = MIN(tds_dstr_len(&(dstr)), 128); \
```
would be better to cap the converted string, not at the source.

ucko replied:
That's fair, but it looks like conversion here is via `tds_convert_stream`, which doesn't accept a length limit.  I suppose it might work to adjust `data_stream.size` directly.

freddy77 replied:
yes, done with also a test

ucko replied:
Great, thanks!  In general, sorry for not adding more tests -- our testing generally takes place via higher-level C++ wrappers.

ucko commented on include/freetds/thread.h:
```diff
@@ -412,6 +412,16 @@ static inline int tds_cond_timedwait(tds_condition *cond, tds_mutex *mtx, int ti
 
 # endif
+#else
+static inline int tds_mutex_noop(void)
```
Fair question.  I don't specifically care about any *platforms* that can't support threads, but a handful of the *configurations* we routinely test effectively use `--disable-threadsafe` and break without these changes.  ("Effectively" because we plug your sources and headers into our own build system.)

ucko commented on src/replacements/getopt.c:
```diff
@@ -39,7 +39,7 @@
 #endif
 
-#if !defined(_WIN32)
+#if !defined(_WIN32)  &&  0
```
Good point.  I'd originally gone with a more readily rebased approach, and evidently didn't fully flip over to a cleaner arrangement.

freddy77 replied:
done

ucko commented on src/tds/sspi.c:
```diff
@@ -282,7 +284,8 @@ tds_sspi_get_auth(TDSSOCKET * tds)
 
-	status = sec_fn->InitializeSecurityContext(&auth->cred, NULL, auth->sname,
+	status = sec_fn->InitializeSecurityContextA
+	       (&auth->cred, NULL, auth->sname,
```
What original encoding should we assume?  UTF-8 nowadays?

ucko commented on win32/initnet.c:
```diff
@@ -4,12 +4,13 @@
 #include <crtdbg.h>
 #endif
 
+#include <winsock2.h>
```
Thanks!

ucko commented on src/odbc/unittests/prepclose.c:
```diff
@@ -43,6 +46,12 @@ close_last_socket(void)
 	if (socketpair(AF_UNIX, SOCK_STREAM, 0, sockets) < 0)
 		return 0;
 
+#if defined(__APPLE__)  &&  defined(SO_NOSIGPIPE)
```
Good idea; the code's pretty simple, but factoring it out would make it easier to accommodate any future platform-specific tuneups that turn out to be necessary.

ucko commented on src/odbc/unittests/common.c:
```diff
@@ -805,6 +809,9 @@ odbc_mark_sockets_opened(void)
 odbc_find_last_socket(void)
 {
+#ifdef TDS_NO_DM
+	return tds_get_s(((TDS_DBC*)odbc_conn)->tds_socket);
```
Great, thanks!

ucko commented on src/tds/login.c:
```diff
@@ -370,10 +370,8 @@ tds_connect(TDSSOCKET * tds, TDSLOGIN * login, int *p_oserr)
-		, 0x703
-		, 0x702
-		, 0x701
+		{ /* 0x704, 0x703, 0x702, */
```
Ah, yeah, this isn't really appropriate upstream as is; sorry it slipped in.  How would you feel about an option to let the client indicate where to start autodetection?

ucko commented on src/tds/login.c:
```diff
@@ -398,12 +396,16 @@ tds_connect(TDSSOCKET * tds, TDSLOGIN * login, int *p_oserr)
 			erc = tds_connect(tds, login, p_oserr);
 			if (TDS_FAILED(erc)) {
 				tds_close_socket(tds);
+				if (tds->conn->env.block_size != orig_size) {
```
Thanks!

ucko commented on src/tds/mem.c:
```diff
@@ -813,7 +813,7 @@ tds_init_login(TDSLOGIN *login, TDSLOCALE * locale)
 	 */
 	login->option_flag2 = TDS_INIT_LANG_REQUIRED|TDS_ODBC_ON;
-	login->tds_version = TDS_DEFAULT_VERSION;
+	login->tds_version = 0 /* was TDS_DEFAULT_VERSION */;
```
Ah, yeah, this change is no longer necessary.

ucko commented on src/utils/threadsafe.c:
```diff
@@ -197,6 +197,9 @@ tds_getservice(const char *name)
 	hints.ai_socktype = SOCK_STREAM;
 	hints.ai_flags = AI_PASSIVE;
+#ifdef AI_CANONNAME
+	hints.ai_flags |= AI_CANONNAME;
```
Good question; this is a very longstanding change, predating my involvement with this code.  I'll check whether we still need it anywhere.

ucko replied:
It certainly doesn't seem to make a difference for either your tests or ours on our current Linux systems; I'll check other platforms, but suspect I can simply drop this change.

ucko commented on include/freetds/thread.h:
```diff
@@ -188,9 +188,9 @@ static inline int tds_raw_cond_wait(tds_condition *cond, tds_raw_mutex *mtx)
+typedef LPTHREAD_START_ROUTINE tds_thread_proc;
 #define TDS_THREAD_PROC_DECLARE(name, arg) \
-	void *WINAPI name(void *arg)
+	DWORD name(void *arg)
```
How so?  I know you'd rather be able to return a pointer here, but on 64-bit Windows, the canonical return type is too narrow, so truncation is a distinct possibility.  I see acknowledging that limitation as the lesser evil, and adjusted all associated code accordingly.

ucko commented on src/tds/data.c:
```diff
@@ -267,6 +267,11 @@ tds_set_param_type(TDSCONNECTION * conn, TDSCOLUMN * curcol, TDS_SERVER_TYPE typ
 		case SYBBINARY:
 			type = XSYBBINARY;
 			break;
+		case SYBBIT:
```
OK, thanks.

ucko commented on src/tds/unittests/common.c:
```diff
@@ -81,7 +89,7 @@ try_tds_login(TDSLOGIN ** login, TDSSOCKET ** tds, const char *appname, int verb
 	if (!tds_set_passwd(*login, PASSWORD)
 	    || !tds_set_user(*login, USER)
-	    || !tds_set_app(*login, appname)
+	    || !tds_set_app(*login, base_appname)
```
Good question.  Perhaps that wasn't part of replacements when I first made this change; I'll be happy to take advantage of it now.

ucko commented on src/tds/mem.c:
```diff
@@ -871,6 +871,7 @@ tds_init_login(TDSLOGIN *login, TDSLOCALE * locale)
 
 	login->valid_configuration = 1;
 	login->check_ssl_hostname = 1;
+	login->use_utf16 = 1;
```
Good point, thanks!

ucko commented on src/ctlib/blk.c:
```diff
@@ -120,6 +120,8 @@ blk_bind(CS_BLKDESC * blkdesc, CS_INT item, CS_DATAFMT * datafmt, CS_VOID * buff
 		colinfo->column_lenbind  = NULL;
 
 		return CS_SUCCEED;
+	} else if (datafmt == NULL) {
```
Thanks!

ucko commented on src/ctlib/blk.c:
```diff
@@ -721,9 +723,10 @@ _blk_get_col_data(TDSBCPINFO *bulk, TDSCOLUMN *bindcol, int offset)
+		    datalen ? *datalen : -1);
 
-	if (*datalen) {
+	if (datalen) {
```
Thanks!

ucko commented on src/tds/challenge.c:
```diff
@@ -163,7 +163,7 @@ static TDSRET
 {
 	const char *user_name, *domain;
-	size_t domain_len, user_name_len, len, buf_usc2le_len = 0;
+	size_t domain_len, user_name_len = 0, len, buf_usc2le_len = 0;
```
Perhaps tagging it with `__attribute__((returns_nonnull))` would help, then.  I agree that warnings aren't necessarily a big deal, but cleaning them up regardless helps any warnings that actually matter stand out more.

ucko commented on src/odbc/unittests/connect.c:
```diff
@@ -11,6 +11,9 @@ init_connect(void)
 }
 
 #ifdef _WIN32
+#ifndef SQL_NOUNICODEMAP
```
I'll take another look, thanks.

ucko replied:
Ah, right... at least as of 1.0, Unicode builds of this test otherwise try to call `SQLGetPrivateProfileStringW` with unsuitable arguments, yielding several warnings -- and a link error when building the ODBC library and tests in driverless mode.

ucko replied:
Good point, `CHKDriverConnect` could otherwise misfire.  It somehow worked for me without even any warnings, but I'll still look into reworking this change (while still accounting for the possibility that `SQLGetPrivateProfileString` should redirect to `tds_SQLGetProfileString`).

ucko commented:
I've rebased my remaining changes against 4c692a0; while I was at it, I also revised some of them to fix issues (mainly memory leaks) GCC 7.3's AddressSanitizer caught and added one new change (ed75d76).  Sorry for the delay, and thanks for cherry-picking a few more fixes in the mean time!

freddy77 commented on src/odbc/unittests/common.c:
```diff
@@ -809,6 +809,9 @@ odbc_mark_sockets_opened(void)
 odbc_find_last_socket(void)
 {
+#ifdef TDS_NO_DM
+	return tds_get_s(((TDS_DBC*)odbc_conn)->tds_socket);
```
This code is already there now, just some couple of lines below with an additional check

ucko replied:
So it is; sorry for missing that!

freddy77 commented on src/dblib/dblib.c:
```diff
@@ -5955,9 +5971,9 @@ dbgetuserdata(DBPROCESS * dbproc)
- * \remarks No effect on behavior of \c db-lib in \c FreeTDS.  
+ * \param version Any DBVERSION_* constant.
+ * \retval SUCCEED if version was valid, FAIL otherwise.
+ * \remarks Use the corresponding protocol version for subsequent connections.
```
I think the purpose of dbsetversion is quite different and should not be changed. Protocol version to connect to a single database can be changed with dbsetlversion/DBSETLVERSION. Setting globally seems a mistake, would prevent multiple connections with different versions.

ucko replied:
FWIW, this patch goes back to the days of Sybase 12.x (or older?), which lacked these APIs.  That said, I'm on board with dropping it at this point.

freddy77 commented:
I'm (very slowly I have to say) picking up changes, testing them and improve/change where needed.

freddy77 commented on src/tds/net.c:
```diff
@@ -669,6 +669,8 @@ tds_select(TDSSOCKET * tds, unsigned tds_sel, int timeout_seconds)
 			switch (sock_errno) {
 			case TDSSOCK_EINTR:
+			case EAGAIN:
+			case TDSSOCK_EINPROGRESS:
```
`pool` does not return these errors as far as I know

ucko replied:
I don't know what Sybase libraries do in this scenario; having retired our last Solaris systems years ago, we now have them only on Linux, which doesn't document either error code as a possible `poll` outcome.  However, both error codes generally call for retrying rather than bailing, and POSIX says `poll` can yield `EAGAIN`: https://www.unix.com/man-page/posix/3p/poll/.  That said, a closer look indicates that the subsequent assertion will need to be updated accordingly; sorry for missing that earlier.

ucko commented on src/ctlib/ct.c:
```diff
@@ -453,7 +453,9 @@ ct_con_props(CS_CONNECTION * con, CS_INT action, CS_INT property, CS_VOID * buff
-			if (*(int *) buffer == CS_TDS_40) {
+			if (*(int *) buffer == 0) {
+				tds_set_version(tds_login, 0, 0);
+			} else if (*(int *) buffer == CS_TDS_40) {
```
Sounds good, thanks!  Thanks also for the latest straight-up merges, and sorry for not getting back to you earlier; I was busy with other matters at the time, but am now revisiting this issue in the course of advancing to 1.4.10; I'll open a fresh PR once I've wrapped that up.

ucko commented on src/odbc/unittests/cancel.c:
```diff
@@ -188,8 +188,12 @@ main(int argc, char **argv)
+		 * has no effect.  (Perhaps the query runs too fast?) */
+		Test(0, 1);
+		Test(1, 1);
+	}
```
Looks good as is now, thanks!

ucko commented on src/tds/data.c:
```diff
@@ -484,7 +484,7 @@ tds_generic_row_len(TDSCOLUMN *col)
 	if (is_blob_col(col))
 		return sizeof(TDSBLOB);
-	return col->column_size;
+	return col->column_size + col->column_varint_size;
```
Also a good question, particularly given that the change accidentally doesn't apply to blobs.  The original version of this change predated some rearrangements that may well have rendered it unnecessary.

ucko commented on src/tds/packet.c:
```diff
@@ -751,10 +754,13 @@ tds_write_packet(TDSSOCKET * tds, unsigned char final)
 	memcpy(tds->out_buf + 8, tds->out_buf + tds->out_buf_max, left);
+	if (res == TDS_SUCCESS	&&  !final  &&	was_final) {
+		tds_write_packet(tds, was_final);
+	}
```
Ah, so it is, thanks!

ucko commented on include/freetds/thread.h:
```diff
@@ -423,6 +423,16 @@ static inline int tds_cond_timedwait(tds_condition *cond, tds_mutex *mtx, int ti
+#  define tds_mutex_init(x) tds_mutex_noop()
+#  define tds_mutex_lock(x) tds_mutex_noop()
+#  define tds_mutex_trylock(x) tds_mutex_noop()
+#  define tds_mutex_unlock(x) do {} while (0)
```
Ah, yeah, at this point, it looks like the only necessary remaining change is to avoid trying to perform extra checks against these dummy replacements.
